[23.2] Provide working routes.url_for every ASGI request

mvdbeek opened the pull request:
Fix the bug portion of https://github.com/galaxyproject/galaxy/issues/17496, though we do of course need to eventually migrate to using starlette's url_for (or handcode urls, as we already do). I think that can wait for the elimination of more makos where this is heavily used.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:

Start up an instance that is configured to submit bug reports and submit one via the API before making any prior requests. All URLs should be correct and not contain `*deprecated attribute, URL not filled in by server*`. 

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

mvdbeek committed:
Provide working routes.url_for every ASGI request

nsoranzo commented on lib/galaxy/webapps/galaxy/api/__init__.py:
```diff
@@ -306,6 +307,16 @@ def get_current_history_from_session(galaxy_session: Optional[model.GalaxySessio
     return None
 
 
+def fix_url_for(mapper, galaxy_request: GalaxyASGIRequest):
```
```suggestion
def fix_url_for(mapper: Mapper, galaxy_request: GalaxyASGIRequest) -> None:
```

mvdbeek replied:
Is this really worth it ? This is legacy stuff we can remove once all relevant routes are FastAPI routes. If we do this I suppose we also need to annotate app.legacy_mapper, which would require typing app more precisely in get_trans.

mvdbeek replied:
It seems to not be an error in mypy, though that then makes me wonder what the point of doing this is in the first place is 😆

nsoranzo replied:
Ah, that's because `getattr(..., ..., None)` returns `Any | None`.
I guess this is still good documentation though.

nsoranzo commented on lib/galaxy/datatypes/genetics.py:
```diff
@@ -114,17 +114,15 @@ def ucsc_links(self, dataset: DatasetProtocol, type: str, app, base_url: str, re
-                    internal_url = "%s" % app.legacy_url_for(
-                        mapper=app.legacy_mapper,
-                        environ=request.environ,
+                    internal_url = "%s" % app.url_for(
```
```suggestion
                    internal_url = app.url_for(
```

nsoranzo commented on lib/galaxy/webapps/galaxy/api/__init__.py:
```diff
@@ -41,6 +41,7 @@
 from fastapi_utils.inferring_router import InferringRouter
 from pydantic import ValidationError
 from pydantic.main import BaseModel
+from routes import request_config
```
```suggestion
from routes import (
    Mapper,
    request_config,
)
```

mvdbeek replied:
I still don't understand why we're doing runtime imports instead of placing this under `if TYPE_CHECKING:`

nsoranzo replied:
I am not against placing them under `if TYPE_CHECKING:`, but I also don't see that providing much of an advantage where there is no risk of a circular import or of adding unnecessary dependencies to a package (which is not the case here).

mvdbeek committed:
Drop unnecessary string substitution

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

mvdbeek committed:
Add type annotation for fix_url_for

I have no clue why this doesn't fail type checking since we don't have
a hint for app.legacy_mapper, but ok.

nsoranzo commented on lib/galaxy/webapps/galaxy/api/__init__.py:
```diff
@@ -306,6 +310,16 @@ def get_current_history_from_session(galaxy_session: Optional[model.GalaxySessio
+    rc = request_config()
+    environ = galaxy_request.environ
+    rc.mapper = mapper
+    rc.environ = environ
```
```suggestion
    rc.environ = galaxy_request.environ
    rc.mapper = mapper
```

mvdbeek committed:
Cut out intermediate environ variable

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>
